ADR/adr-0008-environment-provisioning.md

```diff
@@ -0,0 +1,428 @@
+
+##### Manual Environment Creation - Sandbox
+
+![](../diagrams/ADR-0008/flow-sandbox-manual-creation.jpg)
+
+##### Manual Environment Creation - Cluster
+
+![](../diagrams/ADR-0008/flow-cluster-manual-creation.jpg)
+
+### Mutating DeploymentTargets and Claims
+
+Users may mutate existing [DeploymentTargets] and [DeploymentTargetClaims] in order to, for instance,
+request that their provisioned cluster is scaled up to include more resources. However,
+implementation of resource request changes is provided on a per-provisioner basis. Some may support
+it, and some may not. Most all of our provisioners in the MVP will make no changes to
+a [DeploymentTarget]’s external resources in the event of a resource request change to either the
+[DeploymentTargetClaim] or the [DeploymentTarget].
+
+In the rare case that a provisioner does support resizing external resources - the user should
+request resource changes on the [DeploymentTargetClaim], which should then cause the provisioner to
+resize the external resources modeled by the [DeploymentTarget]. Lastly, the provisioner should update
+the resources in the spec of the [DeploymentTarget] to reflect the external change.
+
+## Consequences
+
+- Better load distribution between the development teams. An addition of a new provisioner type doesn’t require changing any of the existing controllers.
+- Encapsulate the logic of provisioning/deleting a [DeploymentTarget] within the provisioner that is responsible for it.
+- Opens the possibility to create [DeploymentTargetClaims] in advance and by that reducing the waiting time for an ephemeral environment for testing.
+- The design is similar to the design of storage in k8s, so it should look familiar to developers and users.
+- We have a larger API surface than we might ultimately need. Users won’t experience this in the UI, but API-based users will face apparent complexity. The fact that this design emulates the concepts in the k8s storage API should reduce the cognitive load this might impose on users and our engineers.
+
+## Phases of implementation
+
+A addendum about what features can be left out until later iterations.
+
+### Phase 0 (Path to MVP)
+
+- At minimum, create the CRDs and make them available, with no binder or provisioner controllers behind them.
+- Next, modify the Environment CRD and teach the gitops service how to navigate from the linked [DTC] to the [DT] in order to find the Secret that it needs for Argo.
+
+Minimal functionality looks like the BYOC case: No provisioners
+necessary - and even the binder can be omitted in this first pass.
+
+- On workspace initialization, have the NsTemplateTier:
+  - Create Spaces with fixed -dev and -stage suffixes manually as a part of the appstudio tier
+  - Create serviceaccounts with rolebindings to administer the namespaces
+  - Create [DeploymentTargets] to match and reference Secrets that will be created to contains tokens for the serviceaccounts in the previous steps.
```

Comment:
Note, we need to resolve this blocker around Phase 0 here. It doesn't appear to be possible using only static tier templates to get the token from one namespace and make it available as a Secret in another namespace.

Reply:
This is something I was starting to get at above.  It feels like we need something else that's storing the cluster data so we can re-use clusters across multiple workspaces.  For the new-user workspace initialization, we need to copy the information about the Sandbox cluster into the various DeploymentTargets.

Later when a tenant can have multiple workspaces, the workspaces for a single tenant will often share clusters. We shouldn't require users to re-enter BYOC information for every workspace.  We need to be able to define the cluster information once at the tenant level and copy or share that information across multiple workspaces.

Reply:
I assume that a `tenant` equals a user. If so, in order to allow sharing BYOC clusters between workspaces, we will have to associate them with the user that added them to Stonesoup, otherwise, BYOC clusters can leak between workspaces. For example:

Alice is part of workspace A
Bob is part of workspaces, A and B
Alice adds cluster C to workspace A
Can Bob use cluster C in workspace B?

Reply:
For tenants, orgs etc we will have to introduce workspace hierarchy when we can store shared resources like DTC/DT on the parent workspace level.

@ralphbean the problem with creating DTs with a secret reference via tier templates is a pretty big problem. I'm not sure we can work it around without introducing very ugly stonesoup specific hacks in the sandbox core code for MVP.

Reply:
On the problem of creating the secret reference in the appstudio namespace, to me, the choice is between these two:

* Do we invent and accept some hack to make Phase 0 happen?
* Do we ignore Phase 0 and go straight for Phase 1? We can avoid spending time on a temporary hack, but will delay MVP.

To resolve that, we need some estimate of how long the SpaceRequest implementation will take - which looks like the biggest item in Phase 1.

Reply:
OK - an update from conversation today with @alexeykazakov and @MatousJobanek

* What we have right now for the Environment API technically works for static environments for MVP target.
* Argo has a privileged service account that it can use to manage any namespace on the cluster.
* We can defer rolling out this new Environments API until after MVP. We will essentially combine the steps from Phase 0 and Phase 1 into a larger Phase 1.

Reply:
Updated in 5469aeb.